src/main/java/com/sourcegraph/webhook/Dispatcher.java

```diff
@@ -59,36 +72,29 @@ private static Request createRequest(Webhook hook, EventSerializer serializer) {
         return request;
     }
 
-    public static void dispatch(Webhook hook, EventSerializer serializer) {
-        executor.submit(() -> {
-            Request request = createRequest(hook, serializer);
+    @Override
+    public void run() {
+        request.setHeader("X-Attempt-Number", String.valueOf(attempt));
+        try {
+            Response response = (Response) request.executeAndReturn((resp) -> resp);
+            if (response.isSuccessful()) {
+                log.debug("Dispatching webhook (" + serializer.getName() + ") data to URL: [" + hook.endpoint + "] succeeded.");
+                return;
+            }
+        } catch (ResponseException e) {
+            log.debug("Dispatching webhook data (" + serializer.getName() + ") to URL: [" + hook.endpoint + "] failed with error:\n" + e);
+        }
+        attempt++;
 
-            int attempt = 0;
-            while (true) {
-                request.setHeader("X-Attempt-Number", String.valueOf(attempt));
-                try {
-                    Response response = (Response) request.executeAndReturn((resp) -> resp);
-                    if (response.isSuccessful()) {
-                        log.debug("Dispatching webhook (" + serializer.getName() + ") data to URL: [" + hook.endpoint + "] succeeded.");
-                        break;
-                    }
-                } catch (ResponseException e) {
-                    log.debug("Dispatching webhook data (" + serializer.getName() + ") to URL: [" + hook.endpoint + "] failed with error:\n" + e);
-                }
-                attempt++;
+        if (attempt == MAX_ATTEMPTS) {
```

Comment:
I think that conditional should be moved to the beginning of the method. Otherwise you'd send a request AND log that it failed after reaaching max_attempts _in the same method call_.

Reply:
In the current code, this case wouldn't occur due to `Dispatch` only being called once ever. I'm a little hesitant to move this if statement to the top since its purpose is to guard the last line which reschedules the dispatch task from being called past the `MAX_ATTEMPTS`.

If we do move it to the top, it'll reschedule and this will unnecessarily consume a thread which will just return because of the if statement.

I think threads are expensive in java (unlike goroutines in go).

Reply:
I see. I also missed the "return" in the success-case, I think.

But, still, I'd make that message clearer to just say something like "attempts exhausted. won't be scheduling another call".

Reply:
Gotcha! Does this sound better?
`"Exceeded maximum (" + MAX_ATTEMPTS + ") attempts to dispatch webhook data (" + serializer.getName() + ") to URL: [" + hook.endpoint + "]"`

Reply:
Yep!